File: next/src/domain/stream.ts
export type StreamCodec = "h264" | "h265" | "av1" | "unknown";

export interface StreamResolution {
  width: number;
  height: number;
}

export interface StreamVariant {
  name: string;
  resolution?: StreamResolution;
  fps?: number;
  codec?: StreamCodec;
  bitrateKbps?: number;
  aliases: string[];
}

export interface CodecPreference {
  allowed: StreamCodec[];
  preferred?: StreamCodec;
}

export interface QualityConstraints {
  preference: "best" | "worst" | "audioOnly";
  maximumHeight?: number;
  maximumFps?: number;
}

export interface StreamCapabilities {
  variants: StreamVariant[];
  supportsCodecSelection: boolean;
}

function record(value: unknown, name: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new TypeError(`${name} must be an object`);
  }

  return value as Record<string, unknown>;
}

function text(value: unknown, name: string): string {
  if (typeof value !== "string") {
    throw new TypeError(`${name} must be a string`);
  }

  return value;
}

function positiveNumber(value: unknown, name: string): number {
  if (typeof value !== "number" || !Number.isFinite(value) || value <= 0) {
    throw new TypeError(`${name} must be a positive number`);
  }

  return value;
}

export function parseCodec(value: unknown): StreamCodec {
  if (value === "h264" || value === "h265" || value === "av1") {
    return value;
  }

  return "unknown";
}

export function parseStreamCapabilities(value: unknown): StreamCapabilities {
  const input = record(value, "stream capabilities");

  if (!Array.isArray(input.variants)) {
    throw new TypeError("stream capabilities variants must be an array");
  }
  if (typeof input.supportsCodecSelection !== "boolean") {
    throw new TypeError("supportsCodecSelection must be a boolean");
  }

  return {
    variants: input.variants.map((item, index) => {
      const variant = record(item, `variant ${index}`);
      const aliases = variant.aliases;
      if (!Array.isArray(aliases)) {
        throw new TypeError(`variant ${index} aliases must be an array`);
      }

      const parsed: StreamVariant = {
        name: text(variant.name, `variant ${index} name`),
        aliases: aliases.map((alias) => text(alias, `variant ${index} alias`)),
      };

      if (variant.resolution !== undefined) {
        const resolution = record(
          variant.resolution,
          `variant ${index} resolution`,
        );
        parsed.resolution = {
          width: positiveNumber(resolution.width, "resolution width"),
          height: positiveNumber(resolution.height, "resolution height"),
        };
      }
      if (variant.fps !== undefined) {
        parsed.fps = positiveNumber(variant.fps, `variant ${index} fps`);
      }
      if (variant.codec !== undefined) {
        parsed.codec = parseCodec(variant.codec);
      }
      if (variant.bitrateKbps !== undefined) {
        parsed.bitrateKbps = positiveNumber(
          variant.bitrateKbps,
          `variant ${index} bitrateKbps`,
        );
      }

      return parsed;
    }),
    supportsCodecSelection: input.supportsCodecSelection,
  };
}

File: next/src-tauri/src/domain/stream.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum StreamCodec {
    H264,
    H265,
    Av1,
    #[serde(other)]
    Unknown,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamResolution {
    pub width: u32,
    pub height: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamVariant {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolution: Option<StreamResolution>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fps: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub codec: Option<StreamCodec>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bitrate_kbps: Option<u32>,
    pub aliases: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamCapabilities {
    pub variants: Vec<StreamVariant>,
    pub supports_codec_selection: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum QualityPreference {
    Best,
    Worst,
    #[serde(rename = "audioOnly")]
    AudioOnly,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct QualityConstraints {
    pub preference: QualityPreference,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub maximum_height: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub maximum_fps: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CodecPreference {
    pub allowed: Vec<StreamCodec>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub preferred: Option<StreamCodec>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PlayerSettings {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    pub arguments: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Theme {
    System,
    Dark,
    Light,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct NotificationSettings {
    pub live_channels: bool,
    pub playback_errors: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct HotkeySettings {
    pub enabled: bool,
    pub accelerator: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Settings {
    pub schema_version: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub streamlink_path: Option<String>,
    pub quality: QualityConstraints,
    pub codec_preference: CodecPreference,
    pub player: PlayerSettings,
    pub theme: Theme,
    pub language: String,
    pub notifications: NotificationSettings,
    pub hotkey: HotkeySettings,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            schema_version: 1,
            streamlink_path: None,
            quality: QualityConstraints {
                preference: QualityPreference::Best,
                maximum_height: None,
                maximum_fps: None,
            },
            codec_preference: CodecPreference {
                allowed: vec![StreamCodec::H264, StreamCodec::H265, StreamCodec::Av1],
                preferred: None,
            },
            player: PlayerSettings {
                path: None,
                arguments: Vec::new(),
            },
            theme: Theme::System,
            language: "en".into(),
            notifications: NotificationSettings {
                live_channels: false,
                playback_errors: true,
            },
            hotkey: HotkeySettings {
                enabled: false,
                accelerator: "Ctrl+Shift+S".into(),
            },
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TwitchUser {
    pub id: String,
    pub login: String,
    pub display_name: String,
    pub profile_image_url: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "status",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum TwitchSession {
    Anonymous,
    Authenticated {
        user: TwitchUser,
        expires_at: String,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TwitchLoginChallenge {
    pub verification_uri: String,
    pub user_code: String,
    pub expires_in_seconds: u32,
    pub polling_interval_seconds: u32,
}

#[cfg(test)]
mod tests {
    use super::{Settings, StreamCapabilities, TwitchSession};
    use serde::Deserialize;
    use serde_json::{Value, json};

    #[derive(Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct Fixture {
        settings: Value,
        capabilities: Value,
        session: Value,
    }

    fn fixture() -> Fixture {
        serde_json::from_str(include_str!("../../../fixtures/domain-contracts.json"))
            .expect("shared domain fixture must be valid")
    }

    #[test]
    fn shared_fixture_round_trips_in_rust() {
        let fixture = fixture();
        let settings: Settings = serde_json::from_value(fixture.settings.clone()).unwrap();
        let capabilities: StreamCapabilities =
            serde_json::from_value(fixture.capabilities.clone()).unwrap();
        let session: TwitchSession = serde_json::from_value(fixture.session.clone()).unwrap();

        assert_eq!(serde_json::to_value(settings).unwrap(), fixture.settings);
        let serialized_capabilities = serde_json::to_value(&capabilities).unwrap();
        let reparsed_capabilities: StreamCapabilities =
            serde_json::from_value(serialized_capabilities.clone()).unwrap();
        assert_eq!(reparsed_capabilities, capabilities);
        assert_eq!(serialized_capabilities["variants"][0]["fps"], json!(60.0));
        assert!(
            serialized_capabilities
                .get("futureStreamlinkFeature")
                .is_none()
        );
        assert_eq!(serde_json::to_value(session).unwrap(), fixture.session);
    }

    #[test]
    fn mutable_settings_reject_unknown_fields() {
        let mut settings = fixture().settings;
        settings["arbitraryCommand"] = json!("rm -rf /");

        let error = serde_json::from_value::<Settings>(settings).unwrap_err();
        assert!(error.to_string().contains("unknown field"));

        let mut nested_settings = fixture().settings;
        nested_settings["quality"]["futureLimit"] = json!(1);
        let nested_error = serde_json::from_value::<Settings>(nested_settings).unwrap_err();
        assert!(nested_error.to_string().contains("unknown field"));
    }

    #[test]
    fn read_only_capabilities_allow_unknown_fields() {
        let capabilities = serde_json::from_value::<StreamCapabilities>(fixture().capabilities);

        assert!(capabilities.is_ok());
    }
}
